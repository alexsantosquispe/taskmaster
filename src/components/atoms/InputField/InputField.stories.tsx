import type { Meta, StoryObj } from '@storybook/react';

import { WrapperUI } from '@/utils/test.utils';
import { useForm } from 'react-hook-form';
import { InputField } from './InputField';

const meta = {
  title: 'Atoms/InputField',
  component: InputField,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof InputField>;

export default meta;

type Story = StoryObj<typeof InputField>;

export const Default: Story = () => {
  const { control } = useForm();
  const props = {
    label: 'Name',
    name: 'name',
    placeholder: 'Enter your name'
  };

  return (
    <div className="flex gap-4">
      <WrapperUI>
        <InputField {...props} control={control} />
      </WrapperUI>
      <WrapperUI>
        <InputField {...props} isDisabled={true} control={control} />
      </WrapperUI>
      <WrapperUI>
        <InputField {...props} isRequired={true} control={control} />
      </WrapperUI>
      <WrapperUI>
        <InputField
          {...props}
          isRequired={true}
          errorMessage="Error message"
          control={control}
        />
      </WrapperUI>
    </div>
  );
};

Default.args = {};
