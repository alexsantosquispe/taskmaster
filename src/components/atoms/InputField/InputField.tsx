import { INPUT_FIELD_ERROR, INPUT_FIELD_STYLES } from '@/styles';

import type { InputProps } from '@/models/types';
import cn from 'clsx';
import { Controller } from 'react-hook-form';
import { twMerge } from 'tailwind-merge';

export const InputField = ({
  label,
  name,
  control,
  placeholder,
  type = 'text',
  isDisabled,
  errorMessage = '',
  isRequired = false,
  className
}: InputProps) => {
  return (
    <Controller
      control={control}
      name={name}
      render={({ field: { onChange, value } }) => {
        return (
          <div className={twMerge('flex w-full flex-col', className)}>
            <label htmlFor={name} className="mb-0.5 text-sm">
              {label}
              {isRequired && <span className="text-rose-600">&nbsp;*</span>}
            </label>
            <input
              id={name}
              name={name}
              onChange={onChange}
              value={value}
              placeholder={placeholder}
              type={type}
              disabled={isDisabled}
              className={twMerge(
                INPUT_FIELD_STYLES,
                cn({
                  'border-rose-600 dark:border-rose-400': !!errorMessage,
                  'bg-gray-100 placeholder:text-gray-200 hover:cursor-not-allowed dark:bg-white/15 dark:placeholder:text-white/10':
                    isDisabled
                })
              )}
            />
            {errorMessage && (
              <span className={INPUT_FIELD_ERROR}>{errorMessage}</span>
            )}
          </div>
        );
      }}
    />
  );
};
