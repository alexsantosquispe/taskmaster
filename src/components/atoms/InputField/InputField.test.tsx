import '@testing-library/jest-dom';

import { fireEvent, screen } from '@testing-library/react';

import { renderWithForm } from '@/utils/testing/unitTest.util';
import { InputField } from './InputField';

describe('InputField', () => {
  const props = {
    label: 'Name',
    name: 'name',
    placeholder: 'Enter your name'
  };

  describe('styles', () => {
    it('should render the component correctly', () => {
      const component = renderWithForm({
        componentToRender: (control) => {
          return <InputField {...props} control={control} />;
        }
      });

      expect(component).toMatchSnapshot();

      expect(screen.getByText(props.label)).toBeInTheDocument();

      expect(
        screen.getByPlaceholderText(props.placeholder)
      ).toBeInTheDocument();
    });
  });

  describe('behavior', () => {
    const onChangeMock = jest.fn();
    const nameValue = 'John Doe';

    beforeEach(() => {
      jest.clearAllMocks();
    });
    it('should check if the input is disabled', () => {
      renderWithForm({
        componentToRender: (control) => {
          return <InputField {...props} control={control} isDisabled={true} />;
        }
      });

      const input = screen.getByRole('textbox');

      expect(input).toBeInTheDocument();

      expect(input).toBeDisabled();
    });

    it('should call the onChangeText callback', () => {
      renderWithForm({
        componentToRender: (control) => {
          return (
            <InputField
              {...props}
              control={control}
              onChangeText={onChangeMock}
            />
          );
        }
      });

      const input = screen.getByRole('textbox');

      fireEvent.change(input, { target: { value: nameValue } });

      expect(input).toHaveValue(nameValue);

      expect(onChangeMock).toHaveBeenCalledTimes(1);

      expect(onChangeMock).toHaveBeenCalledWith(nameValue);
    });
  });
});
