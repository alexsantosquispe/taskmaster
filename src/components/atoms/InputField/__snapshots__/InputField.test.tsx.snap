// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`InputField styles should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex w-full flex-col"
      >
        <label
          class="mb-0.5 text-sm"
          for="name"
        >
          Name
        </label>
        <input
          class="focus:ring-accent focus:border-accent dark:focus:ring-accent-dark dark:focus:border-accent-dark w-full rounded-lg border border-neutral-200 px-4 py-2 text-sm focus:ring-1 focus:outline-none dark:border-white/20"
          id="name"
          name="name"
          placeholder="Enter your name"
          type="text"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex w-full flex-col"
    >
      <label
        class="mb-0.5 text-sm"
        for="name"
      >
        Name
      </label>
      <input
        class="focus:ring-accent focus:border-accent dark:focus:ring-accent-dark dark:focus:border-accent-dark w-full rounded-lg border border-neutral-200 px-4 py-2 text-sm focus:ring-1 focus:outline-none dark:border-white/20"
        id="name"
        name="name"
        placeholder="Enter your name"
        type="text"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
